Refactor annotation events

LMS007 opened the pull request:
This is mostly just moves 3 parts of functionality from root-thread out into appropriate stores and improves the test cases.

fixes #1236

LMS007 commented on src/sidebar/services/root-thread.js:
```diff
@@ -82,18 +81,6 @@ function RootThread($rootScope, store, searchFilter, viewFilter) {
     });
   }
 
-  function deleteNewAndEmptyAnnotations() {
```
Moved this to drafts store, renamed to `deleteNewAndEmptyDrafts`

LMS007 commented on src/sidebar/services/root-thread.js:
```diff
@@ -111,33 +98,14 @@ function RootThread($rootScope, store, searchFilter, viewFilter) {
   });
 
   $rootScope.$on(events.BEFORE_ANNOTATION_CREATED, function(event, ann) {
-    // When a new annotation is created, remove any existing annotations
```
created a new action called `createAnnotation` in annotations store

LMS007 commented on src/sidebar/services/root-thread.js:
```diff
@@ -111,33 +98,14 @@ function RootThread($rootScope, store, searchFilter, viewFilter) {
   });
 
   // Remove any annotations that are deleted or unloaded
   $rootScope.$on(events.ANNOTATION_DELETED, function(event, annotation) {
```
created new action called `removeAndDeselectAnnotation` in annotations store

LMS007 commented on src/sidebar/services/test/root-thread-test.js:
```diff
@@ -350,38 +359,23 @@ describe('rootThread', function() {
+      );
     });
 
-    unroll(
```
had to break this up since they use different  asserts now

LMS007 commented on src/sidebar/services/test/root-thread-test.js:
```diff
@@ -395,24 +389,6 @@ describe('rootThread', function() {
         fakeStore.state.annotations.push(existingNewAnnot);
       });
 
-      it('removes drafts for new and empty annotations', function() {
```
moves tests to drafts stores

hmstepanek commented on src/sidebar/services/root-thread.js:
```diff
@@ -111,33 +98,14 @@ function RootThread($rootScope, store, searchFilter, viewFilter) {
 
   // Remove any annotations that are deleted or unloaded
   $rootScope.$on(events.ANNOTATION_DELETED, function(event, annotation) {
-    store.removeAnnotations([annotation]);
```
Couldn't this just be store.removeAnnotation(annotation)?

hmstepanek commented on src/sidebar/services/root-thread.js:
```diff
@@ -111,33 +98,14 @@ function RootThread($rootScope, store, searchFilter, viewFilter) {
-    (ann.references || []).forEach(function(parent) {
-      store.setCollapsed(parent, false);
-    });
+    store.createAnnotation(ann);
```
Much cleaner!

hmstepanek commented on src/sidebar/store/modules/annotations.js:
```diff
@@ -331,6 +332,48 @@ function hideAnnotation(id) {
+/**
+ * De-select and remove an annotation.
+ */
+function removeAndDeselectAnnotation(annotation) {
```
My first thought is, looking at the code where this is actually called it appears to always create a list of 1 annotation so I'm not sure there's a need for a plural version anyway. 

Since this is dispatching a singular action inside the plural store function does this really have the advantage of being more efficient? Based on what you describe above it seems like it would still be calling every store subscriber for every dispatched action anyway although I see your point if it was a plural action that didn't dispatch other singular actions then it would be more efficient.

hmstepanek commented on src/sidebar/store/modules/test/drafts-test.js:
```diff
@@ -225,4 +225,40 @@ describe('Drafts Store', () => {
+
+  describe('#deleteNewAndEmptyDrafts', () => {
+    it('should remove only new or empty drafts', () => {
+      const newAnnotations = [
```
Personally I like having the test description as part of the test params for each test case but there are certainly multiple ways to do it!

[{
  description: '...',
  ann: '...'
}].forEach( (test) => {
  it(test.description, () => {
    ...
  });
});

hmstepanek commented on src/sidebar/services/test/root-thread-test.js:
```diff
@@ -341,6 +344,12 @@ describe('rootThread', function() {
   context('when annotation events occur', function() {
     const annot = annotationFixtures.defaultAnnotation();
 
+    it('adds or updates annotations when BEFORE_ANNOTATION_CREATED event occurs', function() {
```
That's fine, I guess my point here was just that I was expecting to see an assertion like addAnnotation or updateAnnotation and instead I saw createAnnotation which doesn't have add or update in the name and then I had to go look at the implementation of createAnnotation to see what it did.

robertknight replied:
Reading this again I get your point. Perhaps "creates a new annotation in the store when `BEFORE_ANNOTATION_CREATED` event occurs" would be clearer.

LMS007 commented:
I'm gonna look over the comments here again, but I think I might need to step back from this PR and form a general approach to this conversion.  It seems to me that there are some cases where an action does not really belong to any one store and then multiple reducers need to handle it. BEFORE_ANNOTATION_CREATE may be one such example.

Also, I'll need to get the functionality correct w/o regression or mistake. Thanks for catching this @robertknight

LMS007 commented:
Added another commit here. There was enough changes that I wanted to show the difference between them so its not a fixup commit. I did some more refactoring, cleaned up confusing/compound actions like removeAndDeselectAnnotation and removeSelectedAnnotation,
cleaned up tests, made them simpler and fixed the bugs that @robertknight discovered.

LMS007 commented on src/sidebar/store/modules/drafts.js:
```diff
@@ -96,14 +97,41 @@ function createDraft(annotation, changes) {
+ */
+
+function deleteNewAndEmptyDrafts() {
+  const annotations = require('./annotations');
```
This one is beyond me, It won't work imported from the top of the file, I could use some help figuring out why. The other stores import just fine.

robertknight replied:
There is a circular import here where the "drafts" module requires the "annotations" module and vice-versa. CommonJS imports don't support circular imports properly. If we converted this to ES imports that would work. Circular imports are often something to be wary of as it indicates coupling which might be problematic in future. In this context, I think there are a few ways this could be done

1. Refactor the logic so that either 1) drafts knows about annotations OR 2) annotations knows about drafts, but not both.
2. Extract this action into a separate module which knows about annotations and drafts
3. Combine the modules together so there is no circular import to worry about

I think I'd prefer to merge this as-is though so we can move on.

robertknight commented on src/sidebar/store/modules/test/drafts-test.js:
```diff
@@ -225,4 +226,44 @@ describe('Drafts Store', () => {
     });
   });
+
+  describe('#deleteNewAndEmptyDrafts2', () => {
```
```suggestion
  describe('#deleteNewAndEmptyDrafts', () => {
```

I think this was a typo

robertknight commented on src/sidebar/store/modules/test/drafts-test.js:
```diff
@@ -225,4 +226,44 @@ describe('Drafts Store', () => {
   });
+
+  describe('#deleteNewAndEmptyDrafts2', () => {
+    unroll(
```
We've been avoiding `unroll` in new code recently in favour of just using an array and `forEach`:

```js
[{
  // Object with properties for test case
}].forEach(({ prop1, prop2, ... }) => {
 it(/* description */, () => {
    // Test case
 });
});
```

The rationale was that not using a custom utility saved developers from having to learn an extra thing, and it made it easier to achieve consistency across projects.

robertknight commented on src/sidebar/store/modules/test/drafts-test.js:
```diff
@@ -225,4 +226,44 @@ describe('Drafts Store', () => {
+            $tag: 'my_annotation_tag',
+          },
+          draft: fixtures.emptyDraft,
+          assertTest: assert.isNotOk,
```
I think it would be preferable here to make the test case property more obviously meaningful, eg. `shouldRemove: true|false` and then use the corresponding assert in the test body.

robertknight commented on src/sidebar/store/modules/test/selection-test.js:
```diff
@@ -1,15 +1,16 @@
 const uiConstants = require('../../../ui-constants');
 
-describe('selection', () => {
+describe('Selection Store', () => {
```
We have a number of different naming schemes for the top-level describe block in tests. This was discussed in Slack a while back before you joined, but I would like us to standardize on just a few:

- If the module exports one function/class, then the name of that function/class is fine providing it is reasonably likely to be unique (eg. `MyReactComponent`)
- If the module has a basename that is reasonably likely to be unique, then that's OK (eg. `build-thread`)
- Otherwise, use the path as if you were importing it from `src` eg. 'store/modules/selection'.

Having more consistent naming should make it easier to know what files to look at when looking at a list of test failures.

robertknight commented on src/sidebar/services/root-thread.js:
```diff
@@ -111,33 +98,14 @@ function RootThread($rootScope, store, searchFilter, viewFilter) {
 
   // Remove any annotations that are deleted or unloaded
   $rootScope.$on(events.ANNOTATION_DELETED, function(event, annotation) {
     store.removeAnnotations([annotation]);
```
Following this refactoring, it looks like there are no remaining listeners for `ANNOTATION_DELETED` or `ANNOTATIONS_UNLOADED`, so a follow-up refactoring can just replace places that emit this event with a direct call to `store.removeAnnotations`.

robertknight commented on src/sidebar/store/modules/drafts.js:
```diff
@@ -96,14 +97,41 @@ function createDraft(annotation, changes) {
+/**
+ * Remove any drafts that are empty.
+ *
+ * An empty draft has no text and no reference tags.
```
Minor nit. We usually refer to the `tags` field of an annotation as just "tags".
